src/apps/chifra/internal/abis/handle_addresses.go

```diff
@@ -0,0 +1,113 @@
+package abisPkg
+
+import (
+	"context"
+	"os"
+	"sort"
+
+	"github.com/TrueBlocks/trueblocks-core/src/apps/chifra/pkg/abi"
+	"github.com/TrueBlocks/trueblocks-core/src/apps/chifra/pkg/account"
+	"github.com/TrueBlocks/trueblocks-core/src/apps/chifra/pkg/logger"
+	"github.com/TrueBlocks/trueblocks-core/src/apps/chifra/pkg/output"
+	"github.com/TrueBlocks/trueblocks-core/src/apps/chifra/pkg/types"
+	"github.com/ethereum/go-ethereum/common"
+)
+
+func (opts *AbisOptions) HandleAddresses() (err error) {
+	testMode := opts.Globals.TestMode
+	result := make(abi.AbiInterfaceMap)
+	if opts.Known {
+		if err = abi.PreloadKnownAbis(opts.Globals.Chain, result, false); err != nil {
+			return
+		}
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	fetchData := func(modelChan chan types.Modeler[types.RawFunction], errorChan chan error) {
+		// result := make(abi.AbiInterfaceMap)
+		for _, addr := range opts.Addrs {
+			address := common.HexToAddress(addr)
+			err = abi.LoadAbiFromAddress(opts.Globals.Chain, address, result)
+			if err == nil {
+				if testMode {
+					// Tests expect sorted map, because maps in C++ are sorted.
+					keys := make([]string, 0, len(result))
+
+					for k := range result {
+						keys = append(keys, k)
+					}
+					sort.Strings(keys)
+
+					for _, k := range keys {
+						modelChan <- result[k]
+					}
+					continue
+				}
+				for _, function := range result {
+					modelChan <- function
+				}
+				continue
+			}
+			if !os.IsNotExist(err) {
+				errorChan <- err
+				cancel()
+			}
+
+			// We didn't find the file
+			// Check if the address is a contract
+			contract, err := account.IsContractAt(
+				opts.Globals.Chain,
+				address,
+				nil, // use latest block number
+			)
+			if err != nil {
+				errorChan <- err
+				cancel()
+			}
+			if !contract {
+				logger.Log(logger.Info, "Address", address, "is not a smart contract. Skipping...")
+				continue
+			}
+			// Fetch ABI from a provider
+			if err = abi.DownloadAbi(opts.Globals.Chain, address, result); err != nil {
+				errorChan <- err
+			}
+			// TODO: This code is duplicated
+			if testMode {
+				// Tests expect sorted map, because maps in C++ are sorted.
```

Comment:
I'm not in love with this...we could add a parameter to DownloadAbi and the above function called `sorted` and it could take a value of `testMode`. Pushes this complication down making this code easier to grok.

Reply:
Yeah, I hate this code, especially duplication. But I felt that by working longer on it I'm blocking articulation, which blocks whole "port to Go" project, so I rushed a bit. But we can (and should) fix it when articulation is done.